algorithms/src/snark/marlin/ahp/indexer/constraint_system.rs

```diff
@@ -27,6 +34,9 @@ pub(crate) struct ConstraintSystem<F: Field> {
     pub(crate) num_public_variables: usize,
     pub(crate) num_private_variables: usize,
     pub(crate) num_constraints: usize,
+    pub(crate) lookup_table: Option<LookupTable<F>>,
+    pub(crate) s_mul: Vec<F>,
+    pub(crate) s_lookup: Vec<F>,
```

Comment:
Instead of `s_mul` and `s_lookup` which are `num_constraints`-sized vectors containing `F`-sized elements, we can save memory as well as generalize to a future with multiple lookup tables as follows:
```rust
...
    pub(crate) mul_constraints: Vec<ConstraintIndex>,
    pub(crate) lookup_constraints: Vec<LookupConstraints<F>>,
}

pub type ConstraintIndex = usize;
pub struct LookupConstraints<F> {
    table: LookupTable<F>,
    indices: Vec<ConstraintIndex>
}
```

If we know we're going to have at most `u32::MAX` constraints, we can even set `ConstraintIndex = u32` to save space further.

Reply:
We can construct the selector polynomials using this compactified information later =)

Reply:
Edited the original comment to change the `Vec<ConstraintIndex>` to a `HashSet<ConstraintIndex>`.